exporters/metric/dogstatsd/internal/statsd/conn.go

```diff
@@ -0,0 +1,291 @@
+)
+
+// NewExporter returns a common implementation for exporters that Export
+// statsd syntax.
+func NewExporter(config Config, adapter Adapter) (*Exporter, error) {
+	if config.MaxPacketSize <= 0 {
+		config.MaxPacketSize = MaxPacketSize
+	}
+	var writer io.Writer
+	var conn net.Conn
+	var err error
+	if config.Writer != nil {
+		writer = config.Writer
+	} else {
+		conn, err = dial(config.URL)
+		if conn != nil {
+			writer = conn
+		}
+	}
+	// TODO: If err != nil, we return it _with_ a valid exporter; the
+	// exporter should attempt to re-dial if it's retryable.  Add a
+	// Start() and Stop() API.
+	return &Exporter{
+		adapter: adapter,
+		config:  config,
+		conn:    conn,
+		writer:  writer,
+	}, err
+}
+
+// dial connects to a statsd service using several common network
+// types.  Presently "udp" and "unix" datagram socket connections are
+// supported.
+func dial(endpoint string) (net.Conn, error) {
+	dest, err := url.Parse(endpoint)
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO: Support tcp destination, need configurable timeouts first.
+
+	scheme := dest.Scheme
+	switch scheme {
+	case "udp", "udp4", "udp6":
+		udpAddr, err := net.ResolveUDPAddr(scheme, dest.Host)
+		locAddr := &net.UDPAddr{}
+		if err != nil {
+			return nil, err
+		}
+		conn, err := net.DialUDP(scheme, locAddr, udpAddr)
+		if err != nil {
+			return nil, err
+		}
+		return conn, err
+	case "unix", "unixgram":
+		scheme = "unixgram"
+		locAddr := &net.UnixAddr{}
+
+		sockAddr, err := net.ResolveUnixAddr(scheme, dest.Path)
+		if err != nil {
+			return nil, err
+		}
+		conn, err := net.DialUnix(scheme, locAddr, sockAddr)
+		if err != nil {
+			return nil, err
+		}
+		return conn, err
+	}
+	return nil, ErrInvalidScheme
+}
+
+// Export is common code for any statsd-based metric.Exporter implementation.
+func (e *Exporter) Export(_ context.Context, checkpointSet export.CheckpointSet) error {
+	buf := &e.buffer
+	buf.Reset()
+
+	var aggErr error
+	var sendErr error
+
+	aggErr = checkpointSet.ForEach(func(rec export.Record) error {
+		before := buf.Len()
+
+		if err := e.formatMetric(rec, buf); err != nil {
+			return err
+		}
+
+		if buf.Len() < e.config.MaxPacketSize {
+			return nil
+		}
+		if before == 0 {
+			// A single metric >= packet size
+			if err := e.send(buf.Bytes()); err != nil && sendErr == nil {
+				sendErr = err
+			}
+			buf.Reset()
+			return nil
+		}
+
+		// Send and copy the leftover
+		if err := e.send(buf.Bytes()[:before]); err != nil && sendErr == nil {
+			sendErr = err
+		}
+
+		leftover := buf.Len() - before
+
+		copy(buf.Bytes()[0:leftover], buf.Bytes()[before:])
+
+		buf.Truncate(leftover)
+		return nil
+	})
+	if err := e.send(buf.Bytes()); err != nil && sendErr == nil {
+		sendErr = err
+	}
+	if sendErr != nil {
+		return sendErr
+	}
+	return aggErr
+}
+
+// send writes a complete buffer to the writer as a blocking call.
+func (e *Exporter) send(buf []byte) error {
+	for len(buf) != 0 {
+		n, err := e.writer.Write(buf)
```

Comment:
specifically I think its this... writer.Write is assumed to accept a buf larger than MTU and write only first MTU bytes... but it doesn't.  it returns an error instead, writing out nothing. I was hacking the code locally before I opted to just wrap the datadog client instead.  I'd rather use this code instead but I think we'll need packet splitting first.

Reply:
I see it. The logic is trying to split packets between each metric record, but the Array-of-points record doesn't support splitting between values.

Reply:
(Fixed.)